pages/design/test-cases.md

```diff
@@ -0,0 +1,70 @@
+
+## Superficial Definition Testing
+
+ACT rules use definitions to avoid duplication, and to "hide" away the details. Definitions such as "semantic role", "visible" and "accessible name" are used in many rules, and are fundamental building blocks of ACT. To avoid having a lot of similar test cases in every rule that uses a particular definition, it is best to keep definition testing to a minimum.
+
+As an example, the "X has non-empty accessible name" rules all have test cases that check for the different ways that an element can have an accessible name. For `img`, that is `alt`, `title`, and `aria-label(ledby)` attributes. That creates sufficient coverage to ensure the basics are understood. But those rules do not go into the details of ensuring accessible name computation is done right. Whether an implementer considers `aria-owns` in the accessible name computation is not relevant for "non-empty accessible name" rules.
+
+That is not to say edge cases in definitions should never have test cases. If there are **real-world examples** of a particular edge case affecting the outcome of a rule, test cases should be added to the rule to cover that edge case.
+
+## Avoid (Accessibility) Support Issues
+
+It is fairly common for rules to involve technologies that are not fully supported. For example, ARIA roles get added every few years, and it takes a while for new roles to get adopted by assistive technologies. In the case that a particular technology has very little or no support at all, the rules should not have test cases for this.
+
+Once there is significant support for a particular technology, it should be considered for test cases. The best way to introduce technologies that are only partially supported is to exclusively provide failed test cases for them. This makes an implementation that does not support this technology partially consistent.
+
+Technologies that are widely supported can be used in all test cases. In all situations, previous principles of deciding if a test case is needed or not still apply.
+
+## Exceptions
+
+It is important to know that there are exceptions to all of these principles. There may be technical reasons not to include a certain test case, or the number of combinations may require an unreasonable number of tests. It may sometimes not make logical sense for some passed test cases to satisfy all accessibility requirements. Some definitions may not require tests, and others can be argued to need extensive tests. Etc.
+
+Letting **real-world** examples drive consistency forward also means rule authors should not get stuck in trying to write test cases for every conceivable situation. Rule authors should strive for a 95% consistency between implementers on first publication. As more examples are discovered, this number will go up over time.
```

Comment:
I think this comes way too late in the document. I think we should be clear from the start: we do not want to test how the different implementations implement technical specifications, such as HTML, CSS, ARIA, or whatever -- we want to test how the different implementations implement the actual issue represented in the ACT Rule in various realistic code examples. I feel this needs to be one of the primary requirements, that all test cases are driven and backed by real-world examples.

PS: I think the XYZ% consistency part is difficult for rule authors to measure and meet, so that such a requirement seems quite unrealistic for us to require.

Reply:
Added an explanation of the percentage.